crates/wasi-common/src/sys/windows/mod.rs

```diff
@@ -1,18 +1,59 @@
 pub(crate) mod clock;
 pub(crate) mod fd;
+pub(crate) mod osdir;
+pub(crate) mod osfile;
 pub(crate) mod oshandle;
+pub(crate) mod osother;
 pub(crate) mod path;
 pub(crate) mod poll;
+pub(crate) mod stdio;
 
+use crate::sys::AsFile;
 use crate::wasi::{types, Errno, Result};
 use std::convert::{TryFrom, TryInto};
 use std::fs::File;
+use std::mem::ManuallyDrop;
+use std::os::windows::prelude::{AsRawHandle, FromRawHandle};
 use std::path::Path;
 use std::time::{SystemTime, UNIX_EPOCH};
 use std::{io, string};
 use winapi::shared::winerror;
 use winx::file::{CreationDisposition, Flags};
 
+impl<T: AsRawHandle> AsFile for T {
+    fn as_file(&self) -> ManuallyDrop<File> {
+        let file = unsafe { File::from_raw_handle(self.as_raw_handle()) };
+        ManuallyDrop::new(file)
+    }
+}
+
+pub(crate) fn get_file_type(file: &File) -> io::Result<types::Filetype> {
+    let file_type = unsafe { winx::file::get_file_type(file.as_raw_handle())? };
+    let file_type = if file_type.is_char() {
+        // character file: LPT device or console
+        // TODO: rule out LPT device
+        types::Filetype::CharacterDevice
+    } else if file_type.is_disk() {
+        // disk file: file, dir or disk device
+        let file = file.as_file();
```

Comment:
*drops in* This is just a `&File -> ManuallyDrop<File>` conversion, right? You ought to be able to remove this line (and the `AsFile` impl above, it seems?

Reply:
Ah, well spotted, thanks!

As far as the `AsFile` impl goes, this blanket impl is needed so that any type that wraps `OsHandle` (which itself implements `AsRawFd`) to be able to convert automatically to `std::fs::File`.

Reply:
That's why, if you look closely, the impl goes for any `T: AsRawFd`.

Reply:
The line removed in 5b7e54e BTW.